kernel-rs/src/lock/spinlock.rs

```diff
@@ -21,6 +23,7 @@ pub struct RawSpinlock {
     ///
     /// Records info about lock acquisition for holding() and debugging.
     locked: AtomicPtr<Cpu>,
+    held: Cell<MaybeUninit<HeldInterrupts>>,
```

Comment:
- locked, held가 꽤 깊은 연관이 있어보입니다: 둘다 non-null/non-uninit이거나, 둘다 null/uninit인 것으로 보이는데 맞나요?
- 혹시 맞다면 위 invariant를 더 잘 표현할 방법이 있을가요? (locked가 atomic이어서 쉬워보이진 않습니다..)
- 애당초 spinlock에서 interrupt off 하는 이유를 혹시 아시나요?

Reply:
* 맞습니다.
* 지금으로서는 딱히 떠오르는 방법이 없는 것 같습니다.
* xv6 문서에 다음과 같이 설명되어 있습니다. (p. 62) 요약하자면, 인터럽트 핸들러에서 spinlock을 acquire하려 할 수 있기 때문에 deadlock을 막으려면 spinlock을 잡은 상태에서는 인터럽트가 꺼져 있어야 합니다.
> The interaction of spinlocks and interrupts raises a potential danger. Suppose `sys_sleep` holds `tickslock`, and its CPU is interrupted by a timer interrupt. `clockintr` would try to acquire `tickslock`, see it was held, and wait for it to be released. In this situation, `tickslock` will never be released: only `sys_sleep` can release it, but `sys_sleep` will not continue running until `clockintr` returns. So the CPU will deadlock, and any code that needs either lock will also freeze.
To avoid this situation, if a spinlock is used by an interrupt handler, a CPU must never hold that lock with interrupts enabled. Xv6 is more conservative: when a CPU acquires any lock, xv6 always disables interrupts on that CPU. Interrupts may still occur on other CPUs, so an interrupt’s acquire can wait for a thread to release a spinlock; just not on the same CPU.